tests/functional/base.py

```diff
@@ -0,0 +1,316 @@
+                self._cleanups.append([self.clients['admin'].openstack,
+                    'role delete', '', '%s' % role])
+
+    def _new_dummy_node(self, owner):
+        node_dir = self.config['dummy_node_dir']
+        if not os.path.exists(node_dir):
+            os.mkdir(node_dir)
+        if not os.path.isdir(node_dir):
+            raise OSError('Error creating dummy node @ %s: not a directory'
+                    % node_dir)
+
+        dummy_node_info = [
+                '{',
+                '   "project_owner_id": "%s",' % 
+                    self.client_info[owner]['project_id'],
+                '   "server_config": {',
+                '       "example_attribute": "example server config",',
+                '       "cpu_type": "Intel Xeon",',
+                '       "cores": 16,',
+                '       "ram_gb": 512,',
+                '       "storage_type": "Samsung SSD",',
+                '       "storage_size_gb": 1024',
+                '   }',
+                '}']
+
+        node = tempfile.mkstemp(prefix='', dir=node_dir, text=True)
+        node_fd = node[0]
+        node_path = node[1]
+        for line in dummy_node_info:
+            os.write(node_fd, ('%s\n' % line).encode())
+        os.close(node_fd)
+
+        self._cleanups.append([os.remove, node_path])
+        return { 'path': node_path, 'uuid': os.path.basename(node_path) }
+
+    def _kwargs_to_flags(self, args):
+        flag_string = ''
+        project_flags = ('owner', 'lessee', 'project')
+        string_flags = ('start_time', 'end_time', 'name', 'resource_type',
+                'resource_uuid', 'status', 'time_range', 'availability_range'
+                'offer_uuid', 'from_owner_id', 'to_owner_id')
+        json_flags = ('properties')
+        standalone_flags = ('long')
+
+        for flag in args.keys():
+            if args[flag] != None:
+                tmp = ' --%s' % flag.replace('_', '-')
+                if flag in string_flags:
+                    flag_string += '%s "%s"' % (tmp, args[flag])
+                elif flag in project_flags:
+                    flag_string += '%s %s' % (tmp,
+                            self.client_info[args[flag]]['project_name'])
+                elif flag in json_flags:
+                    flag_string += '%s "%s"' % (tmp, json.dumps(args[flag]))
+                elif flag in standalone_flags:
+                    flag_string += tmp
+
+        return flag_string
+
+    def _merge_kwargs(self, default, args):
+        for key in default.keys():
+            if key in args.keys():
+                default[key] = args[key]
+        return default
+
+    def parse_details(self, output):
+        listing = output_parser.listing(output)
+        details = {}
+        for item in listing:
+            details.update({ item['Field']: item['Value'] })
+        return details
+
+    def offer_create(self, client, node, parse=True, **kwargs):
```

Comment:
I kinda think that these methods - and therefore the methods above such as `parse_details` and `kwargs_to_flags` - belong in a new file, as they don't really deal with setting up the test base. I know Ironic has these methods in their base, but splitting them out seems much cleaner to me.

Reply:
It definitely would be, but I'm not sure how I'd do that since these methods rely a lot on the reference to `self`.

Reply:
I *think* the methods above don't actually use `self`, and you could just pass in `self.clients[client]`.